pkg/connector/host.go

```diff
@@ -88,18 +96,113 @@ func (r *hostResourceType) List(ctx context.Context, _ *v2.ResourceId, pt *pagin
+		return nil, "", nil, fmt.Errorf("baton-ipa: failed to list hbac rules in '%s': %w", r.baseDN.String(), err)
+	}
+
+	pageToken, err := bag.NextToken(nextPage)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	var rv []*v2.Entitlement
+	for _, hbacRuleEntry := range hbacRuleEntries {
+		accessRule := hbacRuleEntry.GetEqualFoldAttributeValue(attrCommonName)
+		assignmentOptions := []ent.EntitlementOption{
+			ent.WithGrantableTo(resourceTypeUser, resourceTypeGroup),
+			ent.WithDisplayName(fmt.Sprintf("%s Host HBAC Rule %s", resource.DisplayName, accessRule)),
+			ent.WithDescription(fmt.Sprintf("Host-Based Access Control for Host %s via rule '%s'", resource.DisplayName, accessRule)),
+		}
+
+		rv = append(rv, ent.NewAssignmentEntitlement(
+			resource,
+			accessRule,
+			assignmentOptions...,
+		))
+	}
+
+	return rv, pageToken, nil, nil
 }
 
 func (r *hostResourceType) Grants(ctx context.Context, resource *v2.Resource, token *pagination.Token) ([]*v2.Grant, string, annotations.Annotations, error) {
-	return nil, "", nil, nil
+	hostDN := resource.GetExternalId().GetId()
+	if hostDN == "" {
+		return nil, "", nil, fmt.Errorf("baton-ipa: host resource %s has no external ID", resource.DisplayName)
+	}
+
+	bag, page, err := parsePageToken(token.Token, &v2.ResourceId{ResourceType: "hbac_rule"})
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	hbacRuleFilter := fmt.Sprintf(hostHbacRuleFilter, hostDN)
+	hbacRuleEntries, nextPage, err := r.client.LdapSearch(
+		ctx,
+		ldap3.ScopeWholeSubtree,
+		r.baseDN,
+		hbacRuleFilter,
+		nil,
+		page,
+		ResourcesPageSize,
+	)
+	if err != nil {
+		return nil, "", nil, fmt.Errorf("baton-ipa: failed to list hbac rules using filter '%s': %w", hbacRuleFilter, err)
+	}
+
+	pageToken, err := bag.NextToken(nextPage)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	var grants []*v2.Grant
+	for _, hbacRuleEntry := range hbacRuleEntries {
+		accessRule := hbacRuleEntry.GetEqualFoldAttributeValue(attrCommonName)
+		members := parseValues(hbacRuleEntry, []string{attrHBACRuleMemberUser})
+
+		// for each member, lookup the ipaUniqueID and resource type
+		for _, member := range members.ToSlice() {
+			m, err := r.ipaObjectCache.get(ctx, member)
+			if err != nil {
+				return nil, "", nil, fmt.Errorf("baton-ipa: failed to get member %s: %w", member, err)
+			}
+
+			grant, err := newHbacRuleGrantFromDN(resource, accessRule, m.ipaUniqueID, m.resourceType)
+			if err != nil {
+				return nil, "", nil, fmt.Errorf("baton-ipa: failed to create grant for member %s: %w", member, err)
+			}
+
+			grants = append(grants, grant)
+		}
+	}
+
+	return grants, pageToken, nil, nil
 }
 
 func hostBuilder(client *ldap.Client, baseDN *ldap3.DN) *hostResourceType {
 	return &hostResourceType{
-		resourceType: resourceTypeHost,
-		client:       client,
-		baseDN:       baseDN,
+		resourceType:   resourceTypeHost,
+		client:         client,
+		baseDN:         baseDN,
+		ipaObjectCache: newIPAObjectCache(client, baseDN),
```

Comment:
Should this come as a dependency? so we can reuse from the previous cache since we use of 2 Builders